File: game/src/behaviours/selectors.cpp
#include "behaviours/selectors.h"

namespace badlands {

BehaviourResult select_argmax(std::span<const Candidate> candidates, const WorldView& view,
                              const SimFactors& factors) {
    const Candidate* best = nullptr;
    float best_score = 0.0f;  // 0 means "not applicable"; a real winner exceeds it
    for (const Candidate& c : candidates) {
        const float s = c.score(view, factors);
        if (s > best_score) {  // strictly greater -> ties keep the earliest
            best_score = s;
            best = &c;
        }
    }
    return best ? best->act(view, factors) : BehaviourResult{};  // default: Idle in place
}

BehaviourResult select_priority(std::span<const Candidate> candidates, const WorldView& view,
                                const SimFactors& factors) {
    for (const Candidate& c : candidates) {
        if (c.score(view, factors) > 0.0f) {
            return c.act(view, factors);
        }
    }
    return BehaviourResult{};  // nothing applicable -> Idle in place
}

BehaviourResult select_banded(std::span<const ActivityDef> activities,
                              const ActivityWeights& weights, const WorldView& view,
                              const SimFactors& factors) {
    // Bands are tried strictly in order, so the hierarchy is enforced by the
    // CONTROL FLOW rather than by the numbers -- no weight can cross a band.
    for (int32_t b = 0; b < static_cast<int32_t>(ActivityBand::Count); ++b) {
        const auto band = static_cast<ActivityBand>(b);
        const ActivityDef* best = nullptr;
        float best_utility = 0.0f;
        for (const ActivityDef& a : activities) {
            if (a.band != band) {
                continue;
            }
            const float weight = weights.of(a.id);
            if (weight <= 0.0f) {
                continue;  // this actor does not have this activity at all
            }
            const float score = a.score(view, factors);
            if (score <= 0.0f) {
                continue;  // vetoed by its own considerations
            }
            const float utility = weight * score;
            if (utility > best_utility) {  // strictly greater -> ties keep the earliest
                best_utility = utility;
                best = &a;
            }
        }
        if (best != nullptr) {
            return best->act(view, factors);
        }
    }
    return BehaviourResult{};  // nothing applicable anywhere -> Idle in place
}

}  // namespace badlands

File: scripts/brains/nim/selectors.nim
# Port of select_banded (game/src/behaviours/selectors.cpp) -- the only
# selector a hero needs (select_argmax/select_priority drive other archetypes,
# never a hero, so they are not ported). Band loop first (Danger, then
# Normal), then per-band argmax over weight*score with a strict `>` so ties
# keep the earliest list entry -- verbatim, including the early-continue/skip
# order. v2: returns a Suggestion (blocks.nim) instead of v1's Decision.

import abi
import activity_catalog
import hero_view
import blocks

type
  ScoreFn* = proc(v: HeroView, f: BlViewFactors): float32 {.nimcall.}
  ActFn* = proc(v: HeroView, f: BlViewFactors): Suggestion {.nimcall.}

  # Mirrors ActivityDef (behaviours/blocks.h): one row of the activity table.
  ActivityEntry* = object
    id*: int32
    band*: Band
    score*: ScoreFn
    act*: ActFn

proc selectBanded*(activities: openArray[ActivityEntry], weights: array[BL_MAX_ACTIVITIES, float32],
                   v: HeroView, f: BlViewFactors): Suggestion =
  # Bands are tried strictly in order (Danger=0, Normal=1) -- the hierarchy is
  # enforced by control flow, not by the numbers.
  for bandIdx in 0 ..< 2:
    let band = Band(bandIdx)
    var bestIdx = -1
    var bestUtility: float32 = 0.0'f32
    for i in 0 ..< activities.len:
      if activities[i].band != band:
        continue
      let weight = weights[int(activities[i].id)]
      if weight <= 0.0'f32:
        continue  # this actor does not have this activity at all
      let score = activities[i].score(v, f)
      if score <= 0.0'f32:
        continue  # vetoed by its own considerations
      let utility = weight * score
      if utility > bestUtility:  # strictly greater -> ties keep the earliest
        bestUtility = utility
        bestIdx = i
    if bestIdx >= 0:
      return activities[bestIdx].act(v, f)
  # Nothing applicable anywhere -> suggest nothing (BL_INT_NONE), the v2
  # vocabulary's own first-class "no opinion" answer -- a cleaner fit than
  # v1's forced fallback to Idle-in-place ever was. This branch is
  # unreachable for a hero with real factors (Idle's weight is always > 0,
  # so the Normal band always finds at least Idle) -- it fires only when
  # every weight on the wire is exactly 0, which is what src/crates/
  # brainhost's real_hero_wasm_conforms test feeds (an all-zeroed view) and
  # asserts an all-zero-bytes SuggestionWire back (BL_INT_NONE == 0 == the
  # struct's own default), so `result` staying the zero-initialized
  # Suggestion{} is exactly what keeps that acceptance test green.
  discard  # result already default-initialized to Suggestion{} (all zero)
